### PR TITLE
Add ability to disable the colorpicker via reactive or template-driven forms

### PR DESCRIPTION
**Pull request by bradenbiz**

### CODE REVIEW

File: e2e/colorpicker.e2e-spec.ts [hunk 1]
```diff
@@ -62,6 +62,12 @@ describe('Colorpicker', () => {
     });
   }
 
+  async function validateColorpickerDisabled(done: DoneFn): Promise<void> {
+    expect('#screenshot-colorpicker3').toMatchBaselineScreenshot(done, {
+      screenshotName: getScreenshotName('colorpicker-disabled')
+    });
+  }
+
```
Comment (bradenbiz) [inline e2e/colorpicker.e2e-spec.ts:70]: the helper function that validates that the disabled colorpicker looks as it should

File: e2e/colorpicker.e2e-spec.ts [hunk 2]
```diff
@@ -70,6 +76,10 @@ describe('Colorpicker', () => {
     it('should match previous opened screenshot', (done) => {
       validateColorpickerOpened(done);
     });
+
+    it('should match previous disabled screenshot', (done) => {
+      validateColorpickerDisabled(done);
+    });
```
Comment (bradenbiz) [inline e2e/colorpicker.e2e-spec.ts:82]: the new visual test for a disabled colorpicker

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 3]
```diff
@@ -140,8 +140,24 @@ export class SkyColorpickerInputDirective
   @Input()
   public allowTransparency = true;
 
+  @Input()
+  public set disabled(value: boolean) {
+    this._disabled = value;
+    this.skyColorpickerInput.disabled = value;
```
Comment (bradenbiz) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts:146]: set the `sky-colorpicker` component's `disabled` attribute to the set value

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 4]
```diff
@@ -140,8 +140,24 @@ export class SkyColorpickerInputDirective
...
+    if (this.disabled) {
+      this.skyColorpickerInput.backgroundColorForDisplay = '#fff';
+    } else {
+      this.skyColorpickerInput.backgroundColorForDisplay = this.modelValue.hex;
+    }
+  }
```
Comment (bradenbiz) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts:152]: for default theme, this logic is required to properly style the component view

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 5]
```diff
@@ -191,7 +207,10 @@ export class SkyColorpickerInputDirective
       });
 
     this.skyColorpickerInput.updatePickerValues(this.initialColor);
-    this.skyColorpickerInput.backgroundColorForDisplay = this.initialColor;
+
+    if (!this.disabled) {
+      this.skyColorpickerInput.backgroundColorForDisplay = this.initialColor;
+    }
```
Comment (bradenbiz) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts:213]: if the colorpicker component's input is disabled on initialization, do no set the initial value

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 6]
```diff
@@ -263,6 +282,13 @@ export class SkyColorpickerInputDirective
     // Validation
   }
 
+  /**
+   * Implemented as part of ControlValueAccessor.
+   */
+   public setDisabledState(isDisabled: boolean) {
+    this.disabled = isDisabled;
+  }
+
```
Comment (bradenbiz) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts:291]: allows disabling through reactive forms

File: src/app/public/modules/colorpicker/colorpicker.component.html [hunk 7]
```diff
@@ -278,6 +283,10 @@
       'sky-colorpicker-reset-button': 'default',
       'sky-btn sky-btn-icon-borderless sky-colorpicker-reset-button-modern': 'modern'
     }"
+    [disabled]="disabled"
+    [ngClass]="{
+      'isDisabled': disabled
+    }"
```
Comment (bradenbiz) [inline src/app/public/modules/colorpicker/colorpicker.component.html:289]: all of these are for the styling

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable](https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable)

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 8]
```diff
@@ -1151,6 +1159,32 @@ describe('Colorpicker Component', () => {
...
+      fixture.detectChanges();
+
+      component.colorForm.controls['colorModel'].enable();
+
+      await fixture.whenStable();
+      fixture.detectChanges();
+
+      expect(outermostDiv).not.toHaveCssClass('isDisabled');
+    });
+
```
Comment (bradenbiz) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts:1187]: verifies can be enabled and disabled (and that appropriate styles are applied)

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 9]
```diff
@@ -2,26 +2,39 @@
   (themeSettingsChange)="themeSettingsChange($event)"
 ></sky-docs-demo-control-panel-theme>
 
-<div
-  id="screenshot-colorpicker"
+<button
+  class="sky-btn sky-btn-default"
+  id="screenshot-action-button-center-align"
+  type="button"
+  (click)="onToggleAbleColorpicker()"
```
Comment (bradenbiz) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:9]: add button to toggle ability of the first colorpicker

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 10]
```diff
@@ -2,26 +2,39 @@
...
 >
-  <sky-colorpicker
-    labelledBy="FOOBAR"
-    #colorPickerExample1
+  Toggle Colorpicker #1's Ability
+</button>
+
+<form
+  [formGroup]="myForm"
+>
```
Comment (bradenbiz) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:16]: surround colorpicker in form so can work with reactive forms

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 11]
```diff
@@ -2,26 +2,39 @@
...
+        [skyColorpickerInput]="colorPickerExample1"
+        [(ngModel)]="color1"
+      >
+    </sky-colorpicker>
+  </div>
+</form>
 <div
   id="screenshot-colorpicker2"
-  style="height:520px;padding:10px;"
+  style="height:50px;padding:10px;"
```
Comment (bradenbiz) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:37]: modify height of 2nd colorpicker so that 3rd (disabled) colorpicker fits on screen so that visual tests correctly for the disabling functionality

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 12]
```diff
@@ -36,3 +49,22 @@
...
+      [allowTransparency]="false"
+      [disabled]="true"
+      [initialColor]="selectedColor3"
+      [outputFormat]="selectedOutputFormat3"
+      [presetColors]="presetColors3"
+      [skyColorpickerInput]="colorPickerExample3"
+      [(ngModel)]="color3"
+    >
+  </sky-colorpicker>
+</div>
```
Comment (bradenbiz) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:70]: add new disabled colorpicker to run disabled functionality tests on

File: src/app/visual/colorpicker/colorpicker-visual.component.ts [hunk 13]
```diff
@@ -48,12 +55,46 @@ export class ColorpickerVisualComponent {
...
+    '#FFF',
+    '#BD4040',
+    '#617FC2',
+    '#60AC68',
+    '#3486BA',
+    '#E87134',
+    '#DA9C9C',
+    '#A1B1A7',
+    '#68AFEF'
+  ];
```
Comment (bradenbiz) [inline src/app/visual/colorpicker/colorpicker-visual.component.ts:74]: for colorpicker #3

File: src/app/visual/colorpicker/colorpicker-visual.component.ts [hunk 14]
```diff
@@ -48,12 +55,46 @@ export class ColorpickerVisualComponent {
...
   constructor(
+    private formBuilder: FormBuilder,
     private themeSvc: SkyThemeService
-  ) { }
+  ) {}
+
+  public ngOnInit(): void {
+    this.myForm = this.formBuilder.group({
+      colorpicker: new FormControl('')
+    });
```
Comment (bradenbiz) [inline src/app/visual/colorpicker/colorpicker-visual.component.ts:86]: to allow first colorpicker's ability to be toggled

File: src/app/visual/colorpicker/colorpicker-visual.component.ts [hunk 15]
```diff
@@ -48,12 +55,46 @@ export class ColorpickerVisualComponent {
...
     this.themeSvc.setTheme(themeSettings);
   }
 
+  public onToggleAbleColorpicker(): void {
+    if (this.myForm.controls['colorpicker'].disabled) {
+      this.myForm.controls['colorpicker'].enable();
+    } else {
+      this.myForm.controls['colorpicker'].disable();
+    }
+  }
```
Comment (bradenbiz) [inline src/app/visual/colorpicker/colorpicker-visual.component.ts:99]: toggle ability of 1st colorpicker function (triggered by button)

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable](https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable)

File: src/app/public/modules/colorpicker/colorpicker.component.scss [hunk 16]
```diff
@@ -303,11 +326,19 @@
...
+      color: $sky-theme-modern-font-deemphasized-color;
+    }
   }
 
   .sky-colorpicker-reset-button-modern {
     width: 40px;
     height: 40px;
+    &.disabled {
+      background-color: $sky-theme-modern-border-color-neutral-light;
+      border: 1px solid $sky-theme-modern-border-color-neutral-medium;
```
Comment (Blackbaud-ToddRoberts) [inline src/app/public/modules/colorpicker/colorpicker.component.scss:339]: it looks like this border gets doubled up with an inset box-shadow set somewhere else so it ends up being 2px. might not need this if buttons already have the inset box-shadow.

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 17]
```diff
@@ -263,6 +268,19 @@ export class SkyColorpickerInputDirective
     // Validation
   }
 
+  /**
+   * Implemented as part of ControlValueAccessor.
+   */
+   public setDisabledState(isDisabled: boolean) {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts:273]: Code style > your method will need a return type:
```
public setDisabledState(isDisabled: boolean): void {
```

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 18]
```diff
@@ -191,7 +193,10 @@ export class SkyColorpickerInputDirective
       });
 
     this.skyColorpickerInput.updatePickerValues(this.initialColor);
-    this.skyColorpickerInput.backgroundColorForDisplay = this.initialColor;
+
+    if (!this._disabled) {
+      this.skyColorpickerInput.backgroundColorForDisplay = this.initialColor;
+    }
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts:198]: Would this also need to check for the `true` condition as well and set `this.skyColorpickerInput.backgroundColorForDisplay = '#fff';` ?

File: src/app/public/modules/colorpicker/colorpicker.component.html [hunk 19]
```diff
@@ -1,7 +1,8 @@
 <div
   class="sky-input-group"
   [ngClass]="{
-    'sky-colorpicker-hidden': !isVisible
+    'sky-colorpicker-hidden': !isVisible,
+    'disabled': disabled
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker.component.html:5]: We'll need to namespace our classes like so: `sky-colorpicker-disabled`, and `sky-colorpicker-button-disabled` (for the button below).

File: src/app/public/modules/colorpicker/colorpicker.component.html [hunk 20]
```diff
@@ -278,6 +283,10 @@
       'sky-colorpicker-reset-button': 'default',
       'sky-btn sky-btn-icon-borderless sky-colorpicker-reset-button-modern': 'modern'
     }"
+    [disabled]="disabled"
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker.component.html:286]: Code style > check your alphabetical ordering here.

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 21]
```diff
@@ -2,26 +2,39 @@
   (themeSettingsChange)="themeSettingsChange($event)"
 ></sky-docs-demo-control-panel-theme>
 
-<div
-  id="screenshot-colorpicker"
+<button
+  class="sky-btn sky-btn-default"
+  id="screenshot-action-button-center-align"
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:7]: Not that it matters much, but is this the right `id`?

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 22]
```diff
@@ -2,26 +2,39 @@
...
+  type="button"
+  (click)="onToggleAbleColorpicker()"
 >
-  <sky-colorpicker
-    labelledBy="FOOBAR"
-    #colorPickerExample1
+  Toggle Colorpicker #1's Ability
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:11]: Just to be more clear, could we reword this to something like, `Toggle colorpicker disable state` ?

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 23]
```diff
@@ -2,26 +2,39 @@
...
+    id="screenshot-colorpicker"
   >
-    <input
-      [initialColor]="selectedColor1"
-      [outputFormat]="selectedOutputFormat1"
-      [presetColors]="presetColors1"
-      [skyColorpickerInput]="colorPickerExample1"
-      [(ngModel)]="color1"
+    <sky-colorpicker
+      labelledBy="FOOBAR"
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:21]: `labelledBy` pointing to an invalid id. Maybe we could just remove this?

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 24]
```diff
@@ -2,26 +2,39 @@
...
+  (click)="onToggleAbleColorpicker()"
 >
-  <sky-colorpicker
-    labelledBy="FOOBAR"
-    #colorPickerExample1
+  Toggle Colorpicker #1's Ability
+</button>
+
+<form
+  [formGroup]="myForm"
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/colorpicker/colorpicker-visual.component.html:15]: Now that we're adding more types of colorpickers to the visual test page, would you mind adding some headers to provide some clarity? We could do something like selection box with `<h3>`'s: https://github.com/blackbaud/skyux-forms/blob/master/src/app/visual/selection-box/selection-box-visual.component.html#L9

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 25]
```diff
@@ -960,6 +966,33 @@ describe('Colorpicker Component', () => {
...
     }));
 
+    it('should enable and disable AfterViewInit using a template-driven form', async () => {
+
+      let outermostDiv = debugElement.query(By.css('div > sky-colorpicker > div')).nativeElement;
+      fixture.detectChanges();
+
+      expect(outermostDiv).not.toHaveCssClass('disabled');
+
+      fixture.detectChanges();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts:976]: Is this `fixture.detectChanges()` needed? You're not taking any action before it...

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 26]
```diff
@@ -960,6 +966,33 @@ describe('Colorpicker Component', () => {
...
+
+      component.disabled = true;
+
+      fixture.detectChanges();
+      await fixture.whenStable();
+      fixture.detectChanges();
+
+      expect(outermostDiv).toHaveCssClass('disabled');
+
+      fixture.detectChanges();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts:986]: Same question as above.

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 27]
```diff
@@ -1151,6 +1185,32 @@ describe('Colorpicker Component', () => {
...
     }));
 
+    it('should enable and disable AfterViewInit using a reactive form', async () => {
+
+      let outermostDiv = debugElement.query(By.css('form > sky-colorpicker > div')).nativeElement;
+      fixture.detectChanges();
+
+      expect(outermostDiv).not.toHaveCssClass('disabled');
+
+      fixture.detectChanges();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts:1195]: Same question as above.

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 28]
```diff
@@ -1151,6 +1185,32 @@ describe('Colorpicker Component', () => {
...
+
+      component.colorForm.controls['colorModel'].disable();
+
+      await fixture.whenStable();
+      fixture.detectChanges();
+
+      expect(outermostDiv).toHaveCssClass('disabled');
+
+      fixture.detectChanges();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts:1204]: Same question as above.

**Review by Blackbaud-AlexKingman — COMMENTED** [review]
Great job. A few comments for you. Also, I noticed when colorpicker 1 is disabled, I can see the transparent grid background. However, when I disabled colorpicker 3, I can only see the white background. Is that expected?

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable](https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable)

**Comment by bradenbiz** [comment]
> I noticed when colorpicker 1 is disabled, I can see the transparent grid background. However, when I disabled colorpicker 3, I can only see the white background. Is that expected?

i noticed that too. 
image
I have `allowTransparency` set to `false`, which removes the transparent grid.

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 29]
```diff
@@ -140,8 +140,24 @@ export class SkyColorpickerInputDirective
...
+  }
+
+  public get disabled(): boolean {
+    return this._disabled;
+  }
+
   private _initialColor: string;
   private modelValue: SkyColorpickerOutput;
+  private _disabled: boolean;
```
[the remark this answers is not in the document]
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker-input.directive.ts:145]: resolved in last commit

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts (hunk 17 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker-input.directive.ts:273]: resolved in last commit

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts (hunk 18 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker-input.directive.ts:198]: No, because this is only triggered once on initialization, which is actually after `setDisabledState` logic is triggered the first time. 

So all I'm doing here is making sure the `backgroundColorForDisplay` is not reset to the `initialColor` value if the component starts off as disabled

File: src/app/public/modules/colorpicker/colorpicker.component.html (hunk 19 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker.component.html:5]: done in last commit

File: src/app/public/modules/colorpicker/colorpicker.component.scss (hunk 16 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker.component.scss:339]: gotcha. ive set the border to be 0px, so that only the box-sadow remains. (fixed in last commit)

File: src/app/public/modules/colorpicker/colorpicker.component.html (hunk 20 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker.component.html:286]: done in last commit

File: src/app/visual/colorpicker/colorpicker-visual.component.html (hunk 22 above)
Comment (bradenbiz) [reply src/app/visual/colorpicker/colorpicker-visual.component.html:11]: done in last commit

File: src/app/visual/colorpicker/colorpicker-visual.component.html (hunk 23 above)
Comment (bradenbiz) [reply src/app/visual/colorpicker/colorpicker-visual.component.html:21]: done in last commit

File: src/app/visual/colorpicker/colorpicker-visual.component.html (hunk 24 above)
Comment (bradenbiz) [reply src/app/visual/colorpicker/colorpicker-visual.component.html:15]: done in last commit

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 25 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts:976]: fixed in last commit

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 26 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts:986]: fixed in last commit

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 27 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts:1195]: fixed in last commit

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 28 above)
Comment (bradenbiz) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts:1204]: fixed in last commit

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable](https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable)

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable](https://developer.blackbaud.com/skyux-colorpicker?_sp.spa=pr-addAbilityToDisable)

**Review by Blackbaud-ToddRoberts — APPROVED** [review]

**Review by Blackbaud-AlexKingman — APPROVED** [review]
